src/watchman_client.js

```diff
@@ -0,0 +1,341 @@
+'use strict';
+
+var watchman = require('fb-watchman');
+
+/**
+ * Constants
+ */
+
+/**
+ * Singleton that provides a public API for a single connection to a watchman instance for 'sane'.
+ * It tries to abstract/remove as much of the boilerplate processing as necessary
+ * from WatchmanWatchers that use it. In particular, they have no idea whether
+ * we're using 'watch-project' or 'watch', what the 'project root' is when
+ * we internally use watch-project, whether a connection has been lost
+ * and reestablished, etc. Also switched to doing things with promises and known-name
+ * methods in WatchmanWatcher, so as much information as possible can be kept in 
+ * the WatchmanClient, ultimately making this the only object listening directly
+ * to watchman.Client, then forwarding appropriately (via the known-name methods) to 
+ * the relevant WatchmanWatcher(s).
+ *
+ * @class WatchmanClient
+ * @param String dir
+ * @param {Object} opts
+ * @public
+ */
+
+function WatchmanClient() {
+
+  // local vars initialized in getClient
+  var _client;
+  var _clientPromise;
+  var _wildmatch;
+  var _relative_root;
+  var _subscriptionId;
+  var _watcherMap;
+  var _watchmanPath;
+
+  var _clientListeners = null;  // direct listeners from WC to watchman.Client.
+
+  // first time init of local vars. Called again in 'onEnd' if we
+  // are reestablishing the connection.
+  clearLocalVars();
+  
+  // Set up the connection to the watchman client. If the optional
+  // value watchmanPath is passed in AND we're not already in the
+  // middle of setting up the client, use that as the path to the
+  // binary. In either case, return a promise that will be fulfilled
+  // when the client has been set up and capabilities have been returned.
+  function getClient(watchmanPath) {
+    var self = this;
+    
+    if (!_clientPromise) {
+      _clientPromise = new Promise((resolve, reject) => {
+
+        let client = new watchman.Client(watchmanPath ? { watchmanBinaryPath: watchmanPath} : {});
+
+        // reset the relevant local values
+
+        client.capabilityCheck(
+          {optional: ['wildmatch', 'relative_root']},
+          (error, resp) => {
+            if (error) {
+              reject(error);
+            } else {
+              _wildmatch = resp.capabilities.wildmatch;
```

Comment:
Question: do you mean the entire function, or just the part starting at _wildmatch? Since there is already an error response rejection (see line 63 in your snippet) isn't that sufficient? But I do agree that try/catch on the others would be okay in the weird case that resp is invalid (though how it could be I don't know - that would mean that the real watchman client itself is returning bad data.  Comments?